src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -825,14 +836,24 @@ DagState uniqifyState(ImmutableArray<StateForCase> cases, ImmutableDictionary<Bo
                     // Select the next test to do at this state, and compute successor states
                     switch (state.SelectedTest = state.ComputeSelectedTest())
                     {
+                        case BoundDagAssignmentEvaluation e
+                            when state.RemainingValues.TryGetValue(e.Input, out IValueSet? currentValues) |
+                                 state.RemainingValues.TryGetValue(e.Target, out IValueSet? targetValues):
+                            IValueSet commonValues = (currentValues, targetValues) is (IValueSet v1, IValueSet v2)
+                                ? v1.Intersect(v2) : currentValues ?? targetValues ?? throw ExceptionUtilities.Unreachable;
```

Comment:
Can all these branches be hit? I would have thought that `currentValues` couldn't be null here, for example. Consider adding assertions and keeping the code as-is (ie. defensive).

Reply:
>Can all these branches be hit? 

Possibly. The absence of a temp here means either all values are possible or we don't track the remaining values for the type. We don't guard against such things in CheckConsistentDecision when adding the assignment node.

However, if I could prove that the first check is indeed unnecessary (per https://github.com/dotnet/roslyn/pull/53891#discussion_r690938452) we could simplify this a bit.